### ISSUE TITLE
Efficient data marshaling between Haskell/JavaScript

### ISSUE THREAD

**Issue by TerrorJack** [tweag/asterius#43]
Directly related: #23, #40

The current method of moving a piece of data between Haskell(WebAssembly)/JavaScript is both slow and troublesome:

* The set of JSFFI basic types is pretty limited (mainly to make GHC's typechecker happy, we don't have our own yet). So there comes the extra overhead of adding a `StablePtr`/`JSRef` entry whenever moving anything that isn't a simple unboxed value.
* When the marshaling requires traversal of that value, the "driver" code is written in Haskell (e.g. using `foldl'` over a `String`). Even if GHC manages to compile the "driver" to fast worker/wrapper pairs with tight inner loops, there's still overhead.
* Imported JavaScript functions are called a lot; there's overhead in crossing wasm/js boundary, and we'd like to call them only a few times with every conversion.
* The conversion functions aren't currently compiled & tested in the boot libs; they're pasted into every input project as a home module. Seriously!?

When implementing marshaling logic for `bytestring`, the above problems rose again, so time to improve the marshaling story, for `bytestring`, regular `String`s, or any other data structure:

* The interfaces of the data marshaling functions in Haskell should be in a standalone module in a boot lib, and can be `import`ed into any input home module of `ahc-link`. We probably should add something like `asterius-prim` package, similar to `ghcjs-prim`.
* The driver code should really be in JavaScript/WebAssembly, and directly accesses the Haskell heap, abusing knowledge about heap object layouts. Take `JSString -> String` as an example; previously we issue a JavaScript call when moving every code point to Haskell, but we really should just directly build the fully-evaluated Haskell string on the heap and return it.
* [HARD] When passing data into Haskell, we should avoid allocating a `StablePtr`; it's going to be dereferenced and freed soon anyway. This indicates something like `foreign import javascript "f()" f :: SomeHaskellType -> IO SomeHaskellType` is possible.

**Comment by TerrorJack** [comment]
Weird. The attempt to add `asterius-prim` to boot libs fail with cryptic error messages like:

```
src/Asterius/Types.hs:9:14: error:
    Ambiguous occurrence ‘Int’
    It could refer to
       either ‘GHC.Base.Int’,
              imported from ‘GHC.Base’ at src/Asterius/Types.hs:6:1-15
              (and originally defined in ‘GHC.Types’)
           or ‘Asterius.Types.Int’, defined at src/Asterius/Types.hs:9:1
  |
9 | type JSRef = Int
  |              ^^^

src/Asterius/Types.hs:11:17: error:
    Ambiguous occurrence ‘Int’
    It could refer to
       either ‘GHC.Base.Int’,
              imported from ‘GHC.Base’ at src/Asterius/Types.hs:6:1-15
              (and originally defined in ‘GHC.Types’)
           or ‘Asterius.Types.Int’, defined at src/Asterius/Types.hs:9:1
   |
11 | type JSString = JSRef
   |                 ^^^^^
```

Adding `-v` to ghc options showed the mapping of wired-in packages aren't quite right:

```
Configuring asterius-prim-0.0.1...
Glasgow Haskell Compiler, Version 8.7.20181115, stage 2 booted by GHC version 8.6.2
Using binary package database: /home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/package.conf.d/package.cache
package flags []
loading package database /home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/package.conf.d
wired-in package ghc-prim mapped to ghc-prim-0.5.3
wired-in package integer-wired-in mapped to integer-gmp-1.0.2.0
wired-in package base mapped to base-4.12.0.0
wired-in package rts mapped to rts
wired-in package template-haskell mapped to template-haskell-2.15.0.0
wired-in package ghc mapped to ghc-8.7
Created temporary directory: /tmp/ghc22488_0
*** C Compiler:
gcc -fno-stack-protector -x c /tmp/22405-0.c -o /tmp/ghc22488_0/ghc_1.s -no-pie -Wimplicit -S -include /home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/include/ghcversion.h -I/home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/base-4.12.0.0/include -I/home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/integer-gmp-1.0.2.0/include -I/home/terrorjack/.stack/programs/x86_64-linux/ghc-8.7.20181115/lib/ghc-8.7.20181115/include
*** Assembler:
gcc -fno-stack-protector -no-pie -x assembler -c /tmp/ghc22488_0/ghc_1.s -o /tmp/22405-1.o
*** Deleting temp files:
Deleting: /tmp/ghc22488_0/ghc_1.s /tmp/ghc22488_0/ghc_2.rsp
*** Deleting temp dirs:
Deleting: /tmp/ghc22488_0
Preprocessing library for asterius-prim-0.0.1..
Building library for asterius-prim-0.0.1..
[1 of 2] Compiling Asterius.Magic   ( src/Asterius/Magic.hs, /mnt/c/Users/Think/Documents/Haskell/asterius/.stack-work/install/x86_64-linux/ghc-8.7/8.7.20181115/share/x86_64-linux-ghc-8.7.20181115/asterius-0.0.1/.boot/dist/dist/asterius-prim/build/Asterius/Magic.o )
[2 of 2] Compiling Asterius.Types   ( src/Asterius/Types.hs, /mnt/c/Users/Think/Documents/Haskell/asterius/.stack-work/install/x86_64-linux/ghc-8.7/8.7.20181115/share/x86_64-linux-ghc-8.7.20181115/asterius-0.0.1/.boot/dist/dist/asterius-prim/build/Asterius/Types.o )
```

**Comment by TerrorJack** [comment]
All right, the real cause is: `ahc` has baked-in special magic for traversing the AST to handle `JSRef`s and it's conflicted with the way we define `JSRef` in a real module. Easy to fix.

**Comment by TerrorJack** [comment]
Building a big Haskell data structure in only one foreign call without leaping between Haskell/JavaScript is confirmed to work. Remaining work: implement this for arrays, bytestrings, etc, tidy up the `JSRef` interface a bit, then merge back to `master`.

**Comment by TerrorJack** [comment]
One side note which won't be implemented right away, but still worth recording:

I'd really love to add builtin `fromJSON`/`toJSON` operations, so we'll have the convenience of pure JSON values without too much overhead, but we don't have `aeson` and its friends in the boot libs yet. It's possible to include a microscopic port of `aeson` with the following changes:

* In the definition of `Value`, replace `HashMap` with `Map`, `Vector` with `[]`, `Scientific` with `Double`, `Text` with `String`. This cuts down a lot of deps.
* Remove the binary encoder/decoder entirely; we'll rely on `JSON.parse`/`JSON.stringify` provided by the native js engine. The `Parser` stuff used to convert from `Value` to user-defined types is still preserved.
* Add builtin functions to convert between a pure `Value` and a JavaScript value.

**Comment by TerrorJack** [comment]
Fast `ArrayBuffer` -> `ByteString` conversion now works via 9a3d8a9. There are two levels of conversion here:

* `Asterius.Types.fromArrayBuffer` calls an rts builtin routine to directly build an `StgArrBytes` object on the Haskell heap, and return it as a `MutableByteArray# RealWorld`. This can be re-used by `bytestring`, `primitive`, etc.
* `Asterius.ByteString.byteStringFromArrayBuffer` builds a `ByteString` upon a cheap `ForeignPtr` with no finalizers.

**Comment by TerrorJack** [comment]
Interesting. After diving into a bug in 93a602a which resulted in the `MutableByteArray#`'s payload being overwritten by other closures, I figured out that the current way `allocate*` are implemented is unsafe: In normal ghc's sm, the nursery for allocating small closures and the small object pool for `allocate*` functions are distinct; however, in asterius there currently is only one global nursery reused by normal cmm code and `allocate*`, so consider the following case:

```
Main.main1_entry() //  []
         { []
         }
     {offset
       c3vM: // global
           Hp = Hp + 80;
           if (Hp > HpLim) (likely: False) goto c3vS; else goto c3vR;
       c3vS: // global
           HpAlloc = 80;
           R1 = Main.main1_closure;
           call (stg_gc_fun)(R1) args: 8, res: 0, upd: 8;
       c3vR: // global
           (_s3uS::I64) = call "ccall" arg hints:  []  result hints:  [‘signed’] __asterius_jsffi_mainzuMain_2_wrapper();
           (_s3uW::I64) = call "ccall" arg hints:  [‘signed’]  result hints:  [PtrHint] __asterius_fromJSArrayBuffer(_s3uS::I64);
           _s3uX::I64 = _s3uW::I64 + 16;
           _s3uY::I64 = I64[_s3uW::I64 + 8];
           (_s3v2::I64) = call "ccall" arg hints:  [PtrHint,
                                                    ‘signed’]  result hints:  [‘signed’] __asterius_toJSArrayBuffer(_s3uX::I64, _s3uY::I64);
           I64[Hp - 72] = GHC.ForeignPtr.PlainPtr_con_info;
           I64[Hp - 64] = _s3uW::I64;
           _c3vw::P64 = Hp - 69;
           call MO_Touch(_c3vw::P64);
           call "ccall" arg hints:  [‘signed’]  result hints:  [] __asterius_jsffi_mainzuMain_3_wrapper(_s3v2::I64);
           I64[Hp - 56] = sat_s3v9_info;
           P64[Hp - 40] = _c3vw::P64;
           I64[Hp - 32] = _s3uX::I64;
           I64[Hp - 24] = _s3uY::I64;
           I64[Hp - 16] = :_con_info;
           P64[Hp - 8] = GHC.Show.$fShow(,)3_closure;
           P64[Hp] = Hp - 56;
           R4 = GHC.Types.True_closure+2;
           R3 = Hp - 14;
           R2 = GHC.IO.Handle.FD.stdout_closure;
           call GHC.IO.Handle.Text.hPutStr'_entry(R4,
                                                  R3,
                                                  R2) args: 8, res: 0, upd: 8;
     }
 },
```

When the `c3vR` block is entered, the `Hp` is already bumped by `80` and assumed to remain constant before branching out from `main1_entry`; however, the `__asterius_fromJSArrayBuffer` imported function calls into `allocate` of the runtime, which bumps `Hp` again and the subsequent assignments to make heap objects will overwrite the payload.

The solution is very obvious (but take a bit of work): separate the nursery/object pool, and do not touch the nursery in `allocate*` functions.

**Comment by TerrorJack** [comment]
The fix is delivered via 39bc959. I'm leaving this issue open since JSON marshaling isn't implemented yet.

**Comment by TerrorJack** [comment]
Closing this one for the moment. For simple things like strings or buffers, we're already capable of the conversion; for structured data, I think JSON is still the way to go, but we haven't tested `aeson` support for the moment. That will go in its own issue.
